### ISSUE TITLE
Proposal for addition to Relationship field type for queries by Id

### ISSUE THREAD

**Issue by mikehazell** [keystonejs/keystone#482]
# Proposal for addition to Relationship field type for queries by Id

## tl;dr

Add graphql query types to the Relationship field type for matching by id(s). This can dramatically improve response time of relationship queries in cases where the id's are already known.

Using `author` as an example for a **to one** relationship the new type would be `authorId`.

Using `tags` as an example for a **to many** relationship the new types would be `tagsId_some`, `tagsId_none` and `tagsId_every`.

## Use case
I have list of `Posts` with related `authors`. If I want to get a list of posts by author I
currently need to run the following query.

```graphql
{
  allPosts(
    orderBy: "publishedAt_DESC"
    where: {
      author: { id: "5b246e843d4e8f76e7ae2069" }
      publishStatus: published
    }
    first: 4
  ) {
    id
    name
    publishedAt
  }
}
```

Currently the Relationship field type adds a `lookup` to the aggregation. The query looks something like the following.

```js
db.posts.aggregate([
    { '$lookup': {
        from: 'users',
        as: 'cjny6bl380000wwp0ntazepwn_author',
        let: { cjny6bl380000wwp0ntazepwn_author_id: '$author' },
        pipeline: [
            { '$match': { '$expr': { '$eq': [ '$_id', '$$cjny6bl380000wwp0ntazepwn_author_id' ] } } },
            { '$match': { _id: { '$eq': "5b246e843d4e8f76e7ae2069" } } },
            { '$addFields': { id: '$_id' } }
        ]
    } },
    { '$addFields': {
        cjny6bl380000wwp0ntazepwn_author_every: { '$eq': [ { '$size': '$cjny6bl380000wwp0ntazepwn_author' }, 1 ] },
        cjny6bl380000wwp0ntazepwn_author_none: { '$eq': [ { '$size': '$cjny6bl380000wwp0ntazepwn_author' }, 0 ] },
        cjny6bl380000wwp0ntazepwn_author_some: { '$gt': [ { '$size': '$cjny6bl380000wwp0ntazepwn_author' }, 0 ] }
    } },
    { '$match': { '$and': [
        { cjny6bl380000wwp0ntazepwn_author_every: true },
        { publishStatus: { '$eq': 'published' } }
    ] } },
    { '$addFields': { id: '$_id' } },
    { '$sort': { publishedAt: -1 } },
    { '$limit': 4 }
], {});
```

This is really slow taking around 14 seconds to return a response on a reasonably large dataset.

This is Partly because the aggregation framework can't take advantage of the indexes defined on the `post` list given the order of operations. More importantly however if we already know the id of the `user` we don't need to do the lookup at all.

---

## Suggested addition to Relationship type.

The proposal is to add graphql Query Types specifically for searching by id.

For **to one** relationships:

```graphql
    ${path}Id: ID
```

For **to many** relationships:

```graphql
    ${path}Id_some: [ID]
    ${path}Id_none: [ID]
    ${path}Id_every: [ID]
```

In the **to one** case described above, the Graphql query would look like this:

```graphql
{
  allPosts(
    orderBy: "publishedAt_DESC"
    where: {
      authorId: "5b246e843d4e8f76e7ae2069"
      publishStatus: published
    }
    first: 4
  ) {
    id
    name
    publishedAt
  }
}
```

Here, we are changing `author: { id: "ABC" }` to `authorId: "ABC"`.

The final mongo aggregation could then look like this:

```js
db.posts.aggregate([
    { '$match': { '$and': [
        { author: { '$eq': ObjectId('5b246e843d4e8f76e7ae2069') } },
        { publishStatus: { '$eq': 'published' } }
    ] } },
    { '$addFields': { id: '$_id' } },
    { '$sort': { publishedAt: -1 } },
    { '$limit': 4 }
], {})
```

Tests in my current project with appropriate indexes defined have this request responding in 43ms.

**Comment by jesstelford** [comment]
Related: #612

**Comment by molomby** [comment]
@jesstelford Input types and output types are already different (and they should be). Take the password field for example, it takes a `${path}` String but exposes a `${path}_is_set` Boolean. Presumably, a bunch of fields will want to do think kind of thing; exposing things like `${path}_length` and other meta data. 

Alternatively, in this case, should we not add it to the input type too?

I'm in favour of this change generally but can see the attraction to optimising the current types/syntax, as suggested:

> we could do it without needing to introduce any new types - where: { id: "" } is a special case that we can optimize for given we know the way we store the relationship is by just saving the ID

Or maybe even both ways.. I actually prefer the `${path}Id` syntax. 

Regardless this optimisation should be made. It might not be as badly needed on Postgres (guessing, I just trust it to, you know, do "relational" stuff better) but it _will_ be faster there too.
